File: examples/algorithms-python/src/leetcode/lc0072_edit_distance.py
"""LeetCode 72. 编辑距离 — Python 实现

给你两个单词 word1 和 word2，请返回将 word1 转换成 word2 所使用的最少操作数。
你可以对一个单词进行如下三种操作：插入一个字符、删除一个字符、替换一个字符。

对标: LeetCode 72 / docs/13-LeetCode算法面试专题/06-面试专题/03-高频Top100-DP与贪心.md
"""


def min_distance(word1: str, word2: str) -> int:
    """计算将 word1 转换为 word2 的最小编辑距离。

    前置条件 (Pre):
        - word1, word2 长度范围 [0, 500]。

    后置条件 (Post):
        - 返回最小操作数（插入/删除/替换），使得 word1 变为 word2。

    核心思路:
        动态规划：设 dp[i][j] 为 word1[0..i-1] 转换为 word2[0..j-1] 的最小操作数。
        状态转移：
        - 若 word1[i-1] == word2[j-1]：dp[i][j] = dp[i-1][j-1]（无需操作）
        - 否则：dp[i][j] = 1 + min(dp[i-1][j],    # 删除 word1[i-1]
                                   dp[i][j-1],    # 插入 word2[j-1]
                                   dp[i-1][j-1])  # 替换
        初始条件：dp[0][j] = j（全插入），dp[i][0] = i（全删除）。

    复杂度:
        时间复杂度: O(m * n) — m, n 分别为两单词长度。
        空间复杂度: O(n) — 滚动数组优化，仅保留当前行和上一行。

    证明要点:
        - 最优子结构：最优解的最后一步必为三种操作之一。
          若最后操作为删除，则前面部分必为 word1[0..i-2] -> word2[0..j-1] 的最优解。
        - 无后效性：dp[i][j] 仅依赖于左侧、上方、左上方的值。
        - 完备性：所有可能的编辑序列均可由 DP 的决策路径表示。

    Args:
        word1: 源字符串。
        word2: 目标字符串。

    Returns:
        最小编辑距离。
    """
    m, n = len(word1), len(word2)
    # 使用一维数组优化空间
    prev = list(range(n + 1))
    curr = [0] * (n + 1)

    for i in range(1, m + 1):
        curr[0] = i
        for j in range(1, n + 1):
            if word1[i - 1] == word2[j - 1]:
                curr[j] = prev[j - 1]
            else:
                curr[j] = 1 + min(prev[j], curr[j - 1], prev[j - 1])
        prev, curr = curr, prev

    return prev[n]


if __name__ == "__main__":
    # LeetCode 官方示例
    assert min_distance("horse", "ros") == 3, "Example 1 failed"
    assert min_distance("intention", "execution") == 5, "Example 2 failed"

    # 边界条件
    assert min_distance("", "") == 0, "Both empty"
    assert min_distance("abc", "") == 3, "Delete all"
    assert min_distance("", "abc") == 3, "Insert all"
    assert min_distance("abc", "abc") == 0, "Identical"
    assert min_distance("kitten", "sitting") == 3, "Classic example"

    print("All tests passed.")

File: examples/algorithms-rust/src/leetcode/lc0072_edit_distance.rs
/// LeetCode 72. Edit Distance
/// https://leetcode.com/problems/edit-distance/
///
/// Problem: Given two strings word1 and word2, return the minimum number of operations
/// required to convert word1 to word2. Operations: insert, delete, replace.
///
/// Formal specification:
/// - Pre: word1, word2 are ASCII/Unicode strings
/// - Post: result = min ops to transform word1 → word2
///
/// Algorithm: Classic DP with optimal substructure.
/// dp[i][j] = min edit distance between word1[0..i) and word2[0..j).
///
/// Reference: docs/13-LeetCode算法面试专题/02-算法范式专题/08-动态规划.md
/// Lean proof: examples/lean_proofs/leetcode/lc0072_edit_distance.lean

pub fn min_distance(word1: String, word2: String) -> i32 {
    let m = word1.len();
    let n = word2.len();
    let w1: Vec<char> = word1.chars().collect();
    let w2: Vec<char> = word2.chars().collect();

    // dp[j] represents edit distance for word1[0..i) vs word2[0..j)
    let mut dp = vec![0; n + 1];

    // Base case: transform empty word1 to word2[0..j) requires j insertions
    for j in 0..=n {
        dp[j] = j as i32;
    }

    for i in 1..=m {
        let mut prev = dp[0]; // dp[i-1][0]
        dp[0] = i as i32;     // transform word1[0..i) to empty word2: i deletions

        for j in 1..=n {
            let temp = dp[j];
            if w1[i - 1] == w2[j - 1] {
                // Characters match: no operation needed
                dp[j] = prev;
            } else {
                // Choose minimum among delete, insert, substitute
                let delete = dp[j] + 1;      // delete w1[i-1]
                let insert = dp[j - 1] + 1;  // insert w2[j-1]
                let substitute = prev + 1;    // replace w1[i-1] with w2[j-1]
                dp[j] = delete.min(insert).min(substitute);
            }
            prev = temp;
        }
    }

    dp[n]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example_1() {
        assert_eq!(min_distance("horse".to_string(), "ros".to_string()), 3);
    }

    #[test]
    fn test_example_2() {
        assert_eq!(
            min_distance("intention".to_string(), "execution".to_string()),
            5
        );
    }

    #[test]
    fn test_empty_strings() {
        assert_eq!(min_distance("".to_string(), "".to_string()), 0);
        assert_eq!(min_distance("abc".to_string(), "".to_string()), 3);
        assert_eq!(min_distance("".to_string(), "abc".to_string()), 3);
    }

    #[test]
    fn test_same_string() {
        assert_eq!(min_distance("abc".to_string(), "abc".to_string()), 0);
    }

    #[test]
    fn test_single_char() {
        assert_eq!(min_distance("a".to_string(), "b".to_string()), 1);
    }
}
